tools/DeviceExplorer/DeviceExplorer/DevicesProcessor.cs

```diff
@@ -6,78 +6,157 @@
+        /// the device keys and other information to display in the table.
+        /// </remarks>
+        public async Task<List<DeviceEntity>> GetABunchOfDevices()
         {
-            try
+            var listOfDevices = new List<DeviceEntity>();
+            IEnumerable<Device> devices = await registryManager.GetDevicesAsync(maxCountOfDevices);
+            if (devices == null) return listOfDevices;
+
+            foreach (Device device in devices)
             {
-                DeviceEntity deviceEntity;
-                IQuery query = registryManager.CreateQuery("select * from devices", null); ;
+                DeviceEntity deviceEntity = MapDeviceToDeviceEntity(device);
+                listOfDevices.Add(deviceEntity);
+            }
+
+            return listOfDevices;
+        }
 
-                while (query.HasMoreResults)
+        /// <summary>
+        /// This method took well over an hour to load 120,000 devices
+        /// </summary>
+        /// <returns></returns>
+        public async Task GetDeviceSample(
+            int sampleSize, 
+            SortableBindingList<DeviceEntity> list)
+        {
+            try
+            {
+                // Get the first sampleSize of devices, starting at the beginning
+                List<Twin> page = await GetDevicesFromPage(0, sampleSize);
+                foreach (Twin twin in page)
                 {
-                    IEnumerable<Twin> page = await query.GetNextAsTwinAsync();
-                    foreach (Twin twin in page)
-                    {
-                        Device device = await registryManager.GetDeviceAsync(twin.DeviceId);
-                        deviceEntity = new DeviceEntity()
-                        {
-                            Id = twin.DeviceId,
-                            ConnectionState = twin.ConnectionState.ToString(),
-                            LastActivityTime = twin.LastActivityTime,
-                            LastStateUpdatedTime = twin.StatusUpdatedTime,
-                            MessageCount = twin.CloudToDeviceMessageCount,
-                            State = twin.Status.ToString(),
-                            SuspensionReason = twin.StatusReason,
-
-                            ConnectionString = CreateDeviceConnectionString(device),
-                            LastConnectionStateUpdatedTime = device.ConnectionStateUpdatedTime
-                        };
-
-                        deviceEntity.PrimaryThumbPrint = twin.X509Thumbprint?.PrimaryThumbprint;
-                        deviceEntity.SecondaryThumbPrint = twin.X509Thumbprint?.SecondaryThumbprint;
-
-                        deviceEntity.PrimaryKey = device.Authentication?.SymmetricKey?.PrimaryKey;
-                        deviceEntity.SecondaryKey = device.Authentication?.SymmetricKey?.SecondaryKey;
-
-                        listOfDevices.Add(deviceEntity);
-                    }
+                    DeviceEntity deviceEntity = await FetchDeviceFromTwin(twin);
```

Comment:
This should be avoided - only do this _if_ the user asks for the keys. (That would be similar to how the Portal accesses the information.)